Weird filesystem hierarchy with nested modules

**Compiler**: rustc 1.31.0-nightly (14f42a732 2018-10-14)

I'm trying to figure out how module system should work in edition 2018 without `mod.rs` to support it in [IntelliJ Rust](https://github.com/intellij-rust/intellij-rust)

I have the following project structure: https://gist.github.com/Undin/eaef0a4765f5a991f6e62f8d07b878db.
And rustc can't compile this code in edition 2015 and 2018 with the following error
```
error[E0583]: file not found for module `qqq`
 --> src/foo.rs:5:13
  |
5 |     pub mod qqq;
  |             ^^^
  |
  = help: name the file either foo/qqq.rs or foo/qqq/mod.rs inside the directory "src/baz"

error: aborting due to previous error

For more information about this error, try `rustc --explain E0583`.
```
But if I move `qqq.rs` into `src/baz/foo` folder rustc will compile it in both editions.

Problems:
* This code souldn't be compilable in edition 2015.
At least, rustc 1.29.2 shows the following error
```
error[E0658]: mod statements in non-mod.rs files are unstable (see issue #44660)
 --> src/foo.rs:5:13
  |
5 |     pub mod qqq;
  |             ^^^
  |
  = help: on stable builds, rename this file to foo/mod.rs

error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`.
```
* `src/baz/foo/qqq.rs` is rather unexpected because `baz` is child module of `foo`. I think `qqq.rs` should be located in `src/foo/baz` or in `src/baz` folder but I'm not sure.

It would be great to understand how it **should** work to implement the corresponding support in the plugin

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -6426,6 +6426,17 @@ impl<'a> Parser<'a> {
             self.directory.path.to_mut().push(&path.as_str());
             self.directory.ownership = DirectoryOwnership::Owned { relative: None };
         } else {
+            // We have to push on the current module name in the case of relative
+            // paths in order to ensure that any additional module paths from inline
+            // `mod x { ... }` come after the relative extension.
+            //
+            // For example, a `mod z { ... }` inside `x/y.rs` should set the current
+            // directory path to `/x/y/z`, not `/x/z` with a relative offset of `y`.
+            if let DirectoryOwnership::Owned { relative } = &mut self.directory.ownership {
+                if let Some(ident) = relative.take() { // remove the relative offset
+                    self.directory.path.to_mut().push(ident.as_str());
+                }
+            }
             self.directory.path.to_mut().push(&id.as_str());
         }
     }

--- a/src/test/ui/non_modrs_mods_and_inline_mods/non_modrs_mods_and_inline_mods.rs
+++ b/src/test/ui/non_modrs_mods_and_inline_mods/non_modrs_mods_and_inline_mods.rs
@@ -1,0 +1,5 @@
+// compile-pass
+
+mod x;
+
+fn main() {}

--- a/src/test/ui/non_modrs_mods_and_inline_mods/x.rs
+++ b/src/test/ui/non_modrs_mods_and_inline_mods/x.rs
@@ -1,0 +1,5 @@
+// ignore-test: not a test
+
+pub mod y {
+    pub mod z;
+}

--- a/src/test/ui/non_modrs_mods_and_inline_mods/x/y/z/mod.rs
+++ b/src/test/ui/non_modrs_mods_and_inline_mods/x/y/z/mod.rs
@@ -1,0 +1,1 @@
+// ignore-test: not a test

